lib/quicly.c

```diff
@@ -4444,15 +4724,177 @@ static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payloa
+    /* find the CID with smallest sequence number */
+    for (int i = 0; i < QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1; i++) {
+        struct st_quicly_spare_cid_t *c = &conn->super.peer.spare_cids[i];
+        if (!c->is_active)
+            continue;
+
+        if (cid == NULL || cid->sequence > c->sequence) {
+            cid = c;
+        }
+    }
+
+    return cid;
+}
+
+/**
+ * copies NEW_CONNECTION_ID information to an internal storage from a received frame
+ */
+static void store_spare_cid(struct st_quicly_spare_cid_t *spare_cid, const quicly_new_connection_id_frame_t *frame)
+{
+    spare_cid->sequence = frame->sequence;
+    set_cid(&spare_cid->cid, frame->cid);
+    memcpy(spare_cid->stateless_reset_token, frame->stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN);
+    spare_cid->is_active = 1;
+}
+
+/**
+ * compares CIDs, returns zero if identical
+ */
+static int cmp_cid(const quicly_cid_t *l, ptls_iovec_t r)
+{
+    if (l->len != r.len)
+        return 1;
+    return memcmp(l->cid, r.base, l->len);
+}
+
 static int handle_new_connection_id_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
 {
+    int ret;
+    int need_to_replace_cid = 0;
     quicly_new_connection_id_frame_t frame;
-    return quicly_decode_new_connection_id_frame(&state->src, state->end, &frame);
+
+    /* TODO: return error when using zero-length CID */
+
+    if ((ret = quicly_decode_new_connection_id_frame(&state->src, state->end, &frame)) != 0)
+        return ret;
+
+    QUICLY_PROBE(NEW_CONNECTION_ID_RECEIVE, conn, probe_now(), frame.sequence, frame.retire_prior_to);
+
+    if (frame.sequence < conn->super.peer.largest_retire_prior_to) {
+        /* An endpoint that receives a NEW_CONNECTION_ID frame with a sequence number smaller than the Retire Prior To
+         * field of a previously received NEW_CONNECTION_ID frame MUST send a corresponding RETIRE_CONNECTION_ID frame
+         * that retires the newly received connection ID, unless it has already done so for that sequence number. (19.15)
+         * TODO: "unless ..." part may not be properly addressed here (we may already have sent the RCID frame for this
+         * sequence) */
+        schedule_retire_connection_id(conn, frame.sequence);
+        /* do not install this CID */
+        return 0;
+    }
+    /* First, handle retire_prior_to field.
+     * This order is important as it is possible to receive a NEW_CONNECTION_ID frame
+     * such that it retires active_connection_id_limit CIDs and then installs
+     * one new CID.
+     */
+    if (frame.retire_prior_to > conn->super.peer.largest_retire_prior_to) {
+        if (frame.retire_prior_to > conn->super.peer.cid_sequence) {
+            schedule_retire_connection_id(conn, conn->super.peer.cid_sequence);
+            need_to_replace_cid = 1;
+        }
+
+        /* retire CIDs in the spare pool */
+        for (int i = 0; i < QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1; i++) {
```

Comment:
Instead of `QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1`, could we use `sizeof(conn->super.peer.spare_cids) / sizeof(conn->super.peer.spare_cids[0])`?

They are identical, but IMO the latter is better, as this for block would only depend on the definition of `spare_cids`, rather than depending on the definition of `spare_cids` and `QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT`.

Reply:
Makes sense. Alternatively I could define a new constant like `QUICLY_SPARE_CONNECTION_ID_LIMIT` and use it for both definition and usage of `spare_cids[]`, as I personally feel writing down `sizeof(conn->super.peer.spare_cids) / sizeof(conn->super.peer.spare_cids[0])` is a bit annoying.

Reply:
How about creating `quicly_countof` or something alike https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/countof-macro?view=vs-2019?

I agree that using `sizeof` directly is tedious, and IMO that is because we need to write the same expression twice. I think we'd be fine if we could change this to something like `i < quicly_countof(conn->super.peer.spare_cids)`.

I'm not very much in favor of defining another constant that represents the number of elements of the array, because we need to remember that that constant is used as the number of elements of that array.

Reply:
Makes sense. Now that I see #325, I'll use it once it gets merged.